### PR TITLE
Standardize positional datamodule and argument names

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/discussions/7425

1. DataLoader typing
2. Consistent argument names
3. Consistent support for `DataModule` as the second positional argument

I guess these 3 things could be split. Kept them together for convenience.

### Question

Should we change the dataloader arguments to singular instead of plural?

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
I like this, but be aware that it breaking not back-compatible change...

File: pytorch_lightning/utilities/types.py [hunk 1]
```diff
@@ -13,3 +14,9 @@
 _EVALUATE_OUTPUT = List[Dict[str, float]]  # 1 dict per DataLoader
 _PREDICT_OUTPUT = Union[List[Any], List[List[Any]]]
 _PARAMETERS = Iterator[torch.nn.Parameter]
+# TODO: expand allowed train_dataloaders types once fixed.
```
Comment (awaelchli) [inline pytorch_lightning/utilities/types.py:17]: which ones are not listed here? 
a todo for this PR or someone else?

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -668,19 +699,25 @@ def tune(
         self.state.status = TrainerStatus.RUNNING
         self.tuning = True
 
+        if train_dataloader is not None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:702]: good that you check `is not None` rather than direct bool 👍

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -415,22 +414,22 @@ def __init__(
     def fit(
         self,
         model: LightningModule,
-        train_dataloader: Any = None,
-        val_dataloaders: Optional[Union[DataLoader, List[DataLoader]]] = None,
+        train_dataloaders: Optional[Union[TRAIN_DATALOADERS, LightningDataModule]] = None,
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:417]: Not sure about `train_dataloaders` as we support only 1 train_dataloader object. Mind explaining ?

**Review by tchaton — COMMENTED** [review]
LGTM expect train_dataloaders to fit.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:417]: Also, not a huge fan of assigning train_dataloaders to LightingDataModule.

File: pytorch_lightning/utilities/types.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/types.py:17]: Depends on whether #7433 lands first or not

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -463,19 +468,20 @@ def fit(
     def validate(
         self,
         model: Optional[LightningModule] = None,
-        val_dataloaders: Optional[Union[DataLoader, List[DataLoader]]] = None,
+        dataloaders: Optional[Union[EVAL_DATALOADERS, LightningDataModule]] = None,
         ckpt_path: Optional[str] = 'best',
         verbose: bool = True,
         datamodule: Optional[LightningDataModule] = None,
+        val_dataloaders=None,  # noqa TODO: remove with 1.6
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:475]: val_dataloaders is only kept for backwards compatibility until 1.6

**Comment by carmocca** [comment]
> but be aware that it breaking not back-compatible change...

I believe I kept the compatibility - with things properly deprecated. What's missing?

**Review by Borda — COMMENTED** [review]
consider using this wrapper https://github.com/Borda/pyDeprecate#self-argument-mapping
with mapping to rank zero https://github.com/PyTorchLightning/pytorch-lightning/blob/96433d03ea6323adea9a822a8222f7ea7ffba99b/pytorch_lightning/metrics/utils.py#L32

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]

File: docs/source/advanced/multiple_loaders.rst [hunk 5]
```diff
@@ -91,23 +91,6 @@ For more details please have a look at :paramref:`~pytorch_lightning.trainer.tra
 Furthermore, Lightning also supports that nested lists and dicts (or a combination) can
 be returned.
 
-.. testcode::
-
-    class LitModel(LightningModule):
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/advanced/multiple_loaders.rst:96]: It's pretty much duplicated. The multiple dataloader docs need an overhaul
